File: onyx-database/src/main/java/com/onyx/diskmap/base/concurrent/DefaultAtomicCounter.java
package com.onyx.diskmap.base.concurrent;

import com.onyx.buffer.BufferStream;
import com.onyx.buffer.BufferStreamable;
import com.onyx.exception.BufferingException;
import com.onyx.persistence.context.SchemaContext;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.atomic.AtomicLong;

/**
 * Created by tosborn1 on 3/2/17.
 * <p>
 * Default iplementation of the atomic counter
 *
 * @since 1.3.0
 */
public class DefaultAtomicCounter implements AtomicCounter, BufferStreamable {

    private AtomicLong aLong;

    /**
     * Constructor
     */
    @SuppressWarnings("unused")
    public DefaultAtomicCounter() {
        aLong = new AtomicLong(0);
    }

    /**
     * Constructor with initial value
     *
     * @param initialValue value to set counter to
     */
    public DefaultAtomicCounter(@SuppressWarnings("SameParameterValue") long initialValue) {
        this.aLong = new AtomicLong(initialValue);
    }

    /**
     * Set long value
     *
     * @param value count
     */
    @Override
    public void set(long value) {
        this.aLong.set(value);
    }

    /**
     * Get counter value
     *
     * @return current value
     */
    @Override
    public long get() {
        return this.aLong.get();
    }

    /**
     * Add value and add more
     *
     * @param more How many more bytes
     * @return The current value
     */
    @Override
    public long getAndAdd(int more) {
        return this.aLong.getAndAdd(more);
    }

    /**
     * Read from buffer
     *
     * @param buffer Buffer Stream to read from
     * @throws BufferingException General exception
     */
    @Override
    public void read(BufferStream buffer) throws BufferingException {
        this.aLong = new AtomicLong(buffer.getLong());
    }

    /**
     * Write to a buffer
     *
     * @param buffer Buffer IO Stream to write to
     * @throws BufferingException cannot write
     */
    @Override
    public void write(BufferStream buffer) throws BufferingException {
        buffer.putLong(aLong.get());
    }

    @Override
    public void write(@NotNull BufferStream buffer, @Nullable SchemaContext context) throws BufferingException {
        write(buffer);
    }

    @Override
    public void read(@NotNull BufferStream buffer, @Nullable SchemaContext context) throws BufferingException {
        read(buffer);
    }
}

File: onyx-database/src/main/kotlin/com/onyx/lang/concurrent/impl/DefaultAtomicCounter.kt
package com.onyx.lang.concurrent.impl

import com.onyx.buffer.BufferStream
import com.onyx.buffer.BufferStreamable
import com.onyx.lang.concurrent.AtomicCounter
import com.onyx.exception.BufferingException
import com.onyx.persistence.context.SchemaContext

import java.util.concurrent.atomic.AtomicLong

/**
 * Created by Tim Osborn on 3/2/17.
 *
 *
 * Default implementation of the atomic counter
 *
 * @since 1.3.0
 */
class DefaultAtomicCounter(initialValue: Long = 0L) : AtomicCounter, BufferStreamable {

    private var aLong: AtomicLong = AtomicLong(initialValue)

    /**
     * Set long value
     *
     * @param value count
     */
    override fun set(value: Long) = this.aLong.set(value)

    /**
     * Get counter value
     *
     * @return current value
     */
    override fun get(): Long = this.aLong.get()

    /**
     * Add value and add more
     *
     * @param more How many more bytes
     * @return The current value
     */
    override fun getAndAdd(more: Int): Long = this.aLong.getAndAdd(more.toLong())

    /**
     * Read from buffer
     *
     * @param buffer Buffer Stream to read from
     * @throws BufferingException General exception
     */
    @Throws(BufferingException::class)
    override fun read(buffer: BufferStream) {
        this.aLong = AtomicLong(buffer.long)
    }

    /**
     * Write to a buffer
     *
     * @param buffer Buffer IO Stream to write to
     * @throws BufferingException cannot write
     */
    @Throws(BufferingException::class)
    override fun write(buffer: BufferStream) = buffer.putLong(aLong.get())

    /**
     * Read from buffer
     *
     * @param buffer Buffer Stream to read from
     * @param context Schema Context
     * @throws BufferingException General exception
     */
    @Throws(BufferingException::class)
    override fun read(buffer: BufferStream, context: SchemaContext?) {
        this.aLong = AtomicLong(buffer.long)
    }

    /**
     * Write to a buffer
     *
     * @param buffer Buffer IO Stream to write to
     * @param context Schema Context
     * @throws BufferingException cannot write
     */
    @Throws(BufferingException::class)
    override fun write(buffer: BufferStream, context: SchemaContext?) = buffer.putLong(aLong.get())

}
